Fix race conditions in Close/Flush functions

lzap opened the pull request:
The `sinit` package provides a function called `Flush` that is supposed to be called from the `main` function (via `defer`). What it does in fact is it closes the writers/handlers. This patch fixes that, it should have really been named `Close` from the getgo.

Now, @mvo5 found out (https://github.com/osbuild/logging/pull/29) found out there is a race condition when calling `Flush` function. This patch fixes that too, the problem is that writers are asynchronous and `Flush` call is implemented as an empty message, but `Close` actually closes the channel to ensure no data can be written during close operation so any incoming logs would panic. This is a correct behavior when shutting down such component I believe.

The solution is to introduce `mutex` and use it for both `Close` and `Flush` methods. I did the same for the `cloudwatchwriter2` which is a port of another open source project but for AWS SDK v2 we used for `launch` and `logging` library now uses it for CloudWatch logging.

In addition, the newly introduced `Close` method now use `sync.Once` to ensure it can only be called once. On top of that, it requires `time.Duration` argument which is then divided by three and each and every `Close` method is called with duration to match the total waiting. Previously this was hardcoded to 6 seconds (3*2 seconds). This is very useful in OpenShift / K8s when pod is shutting down (it is terminated after 30 seconds).

This PR comes in multiple commits. Thanks @mvo5 for finding the issue.

lzap commented on pkg/sinit/flush.go:
```diff
@@ -0,0 +1,61 @@
+	result := true
+
+	onceClose.Do(func() {
+		each := timeout / 3
```
Genius! :-)

lzap commented on pkg/splunk/client.go:
```diff
@@ -219,27 +221,44 @@ func (sl *splunkLogger) sendPayloads(buf *bytes.Buffer) error {
-	timeout := time.Now().Add(2 * time.Second)
-	for sl.active.Load() {
-		time.Sleep(10 * time.Millisecond)
+		timeout := time.Now().Add(timeout)
```
That is not guaranteed, I am adding a mutex to the main Close method, therefore it can be blocked for some time during concurrent load.

lzap replied:
Oh sure I thought this is the sinit package but indeed this is the splunk client.

lzap commented on pkg/splunk/client.go:
```diff
@@ -219,27 +221,44 @@ func (sl *splunkLogger) sendPayloads(buf *bytes.Buffer) error {
+
+		close(sl.payloads)
 
-	timeout := time.Now().Add(2 * time.Second)
```
Yes this looks cleaner, but the patch is getting quite big, created https://github.com/osbuild/logging/issues/33

lzap commented on pkg/sinit/init.go:
```diff
@@ -145,153 +145,164 @@ type LogrusConfig struct {
+	if res != nil {
+		return ErrAlreadyInitialized
+	}
+	res = &resources{}
```
Nothing was changed in the "once" block after this line, the indentation change hunk is a bit hard to read.

mvo5 commented on pkg/sinit/flush_test.go:
```diff
@@ -0,0 +1,49 @@
+
+	go Close(50 * time.Millisecond)
+	go Close(50 * time.Millisecond)
+}
```
Sorry, I should have been more clear. We have a function called "Close()" but it does not close e.g. the StdoutConfig. The question is if this is a bug or we should document it or if we ignore it.

mvo5 replied:
Any thoughts on this question?

lzap replied:
Oh right, missed this, added this:

```
	// Close standard logger
	if res.prevSlogger != nil {
		slog.SetDefault(res.prevSlogger)
		res.prevSlogger = nil
	}

	// Close strc logger
	strc.SetNoopLogger()

	// Close logrus logger
	logrus.SetDefault(logrus.NewDiscardProxy())
```

lzap replied:
Created an issue: https://github.com/osbuild/logging/issues/35

mvo5 reviewed (dismissed):
Thanks for fixing this race and adding the tests. I like it. I have some inline suggestions but no blockers. Might be worth to add the one test though (could a separate PR of course too as AFAICT this time-out is pre-existing).

lzap commented on pkg/sinit/flush.go:
```diff
@@ -0,0 +1,123 @@
+	// Allow re-initialization
+	res = nil
+
+	// If the current time is after the timeout, add ErrTimeoutDuringClose to the error list
```
Because sentry SDK does not return errors and I wanted to have a common error that the caller can check against.

lzap replied:
Because sentry SDK does not return errors and I wanted to have a common error that the caller can check against.

lzap commented:
Amended all your remarks, including cleaning up the error joining it is now much nicer.

Added the test you recommend without changes just I had to add `atomic` integer because race detector was not happy without it, also dropped the comment about GH CI, I do not think this is worth embedding in the codebase (we might migrate somewhere else to something faster one day).

mvo5 commented on pkg/splunk/client_test.go:
```diff
@@ -190,3 +193,37 @@ func TestSplunkLoggerPayloads(t *testing.T) {
+
+	err = sl.close(closeTimeout)
+	assert.Equal(t, ErrCloseTimeout, err)
+	assert.True(t, time.Since(start) < 5*closeTimeout)
```
Having some comment here (like `// add a lot of slack to the closeTimeout to account for slow  GH VMs that may run the tests`  would be nice I think, just to make sure there is no wondering why we are not checking more exactly for the timeout.

mvo5 reviewed (approved):
I like it! one tiny suggestion about a comment and the question about how much close should close inline but not blockers (but I think the comment would be nice).

lzap commented:
I do not see the question, I cannot even reply, GH is ill...

<img width="1010" height="317" alt="image" src="https://github.com/user-attachments/assets/7e19641c-58c6-4133-b685-c9d1a9ccd74b" />

lzap commented:
Resolved both comments.
